core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/entity/AbstractEntitySelector.java

```diff
@@ -10,4 +12,30 @@
 public abstract class AbstractEntitySelector<Solution_> extends AbstractSelector<Solution_>
         implements EntitySelector<Solution_> {
 
+    /**
+     * Entity selectors need to implement equality,
+     * so that things like pillar cache can work properly.
+     * In order to enforce the correct equality checks on all entity selectors,
+     * they are required to override this method.
+     * {@link #equals(Object)} and {@link #hashCode()} are then implemented here, and made final.
+     *
+     * @return never null, objects to be used as basis for the equality check
+     */
+    protected abstract Object[] getEqualityRequirements();
```

Comment:
In order to be able to tell which pillar selectors can share pillar cache, I need entity selectors to support equality.
This requirement forces me to propagate equality checks to a lot of places.

Reply:
Why not just force them to implement equals and hashcode?